copilot: Improve sign-in flow

KushalMeghani1644 opened the pull request:
This PR significantly improves the GitHub Copilot sign-in experience in Zed by refactoring the code architecture and enhancing the user interface.

### Key Changes
- **🏗️ Architecture**: Extracted business logic into `CopilotSignInService` for better separation of concerns
- **🔧 State Management**: Replaced simple boolean with `ConnectionState` enum for more robust state handling
- **🛡️ Type Safety**: Added type-safe wrappers (`UserCode`, `VerificationUri`) and comprehensive error handling
- **🎨 UX Improvements**: Enhanced UI with visual feedback, better loading states, and improved accessibility
- **🧪 Testing**: Added basic unit tests for validation logic

### Technical Details
- Modularized modal rendering with dedicated methods for each state
- Implemented proper error propagation with custom `SignInError` type
- Added toast management helper for consistent user notifications
- Improved clipboard handling and device code validation
- Enhanced visual design with icons, colors, and responsive layout

### Testing
- [ ] Sign-in flow works correctly for new users
- [ ] Reinstall and sign-in process functions properly
- [ ] Error states display appropriate messages and recovery options
- [ ] Toast notifications appear and dismiss correctly
- [ ] Modal accessibility and keyboard navigation work as expected

KushalMeghani1644 committed:
refactor(copilot): improve sign-in flow with better error handling and UX

- Extract business logic into service layer for better separation of concerns
- Replace boolean state with ConnectionState enum for robust state management
- Add type-safe wrappers and comprehensive error handling
- Enhance UI with visual feedback, loading states, and accessibility improvements

KushalMeghani1644 committed:
Fixed compilation errors

cyberphantom52 commented on crates/copilot/src/sign_in.rs:
```diff
@@ -10,44 +10,171 @@ use util::ResultExt as _;
+#[derive(Debug, Clone)]
+struct UserCode(String);
+
+impl UserCode {
+    fn as_str(&self) -> &str {
+        &self.0
+    }
+}
```
maybe this can be replaced with `impl Deref`?

KushalMeghani1644 committed:
fix(copilot): update sign-in modal to use current GPUI APIs

- Add missing imports (Deref, actions macro, rems function)
- Update spawn calls to use new async closure syntax
- Fix clipboard API calls (read_from_clipboard vs try_read_from_clipboard)
- Fix open_url calls to use window instead of context
- Replace menu::Cancel with local Cancel action
- Improve error handling and remove problematic return statement
- Format code and improve structure for better readability

KushalMeghani1644 committed:
Fixed compilation errors

notpeter reviewed (commented):
This is clearly a PR written by AI which you haven't bothered to remove the WIP comments or complete the implementation.

In the future:
1. Review your code before asking us to review it.
2. Do not provide an AI-slop generated PR description.

Happy to review a future clean PR.
Thanks
